src/Conversion/ONNXToKrnl/Tensor/Unique.cpp

```diff
@@ -0,0 +1,194 @@
+/// input options.
+Value emitArgUnique(ConversionPatternRewriter &rewriter, Location loc,
+    Value total, Value input, int64_t axis, int64_t sorted, Value Y,
+    Value indices, Value inverse_indices, Value counts, bool count_only) {
+  MultiDialectBuilder<KrnlBuilder, IndexExprBuilderForKrnl, MathBuilder> create(
+      rewriter, loc);
+  IndexExprScope scope(create.krnl);
+  MemRefType inputMemRefType = input.getType().cast<MemRefType>();
+  int64_t rank = inputMemRefType.getRank();
+  assert(axis < rank && "axis is out of bound");
+  LiteralIndexExpr zeroIE(0), oneIE(1);
+  SmallVector<IndexExpr, 4> lbs(rank, zeroIE);
+  SmallVector<IndexExpr, 4> ubs;
+  create.krnlIE.getShapeAsDims(input, ubs);
+
+  // Emit krnl.Call to call omTensorUnique API
+  Type int_type = rewriter.getIntegerType(64);
+  Value val_axis = create.math.constant(int_type, axis);
+  Value val_sorted = create.math.constant(int_type, sorted);
+  if (count_only) {
+    SmallVector<Value, 4> operands = {total, input, val_axis, val_sorted};
+    rewriter.create<KrnlCallOp>(loc, "omTensorUniqueCount", 1, operands);
+  } else {
+    SmallVector<Value, 8> operands = {total, input, val_axis, val_sorted, Y,
+        indices, inverse_indices, counts};
+    rewriter.create<KrnlCallOp>(loc, "omTensorUnique", 8, operands);
+  }
+  return total;
+}
+
+struct ONNXUniqueOpLowering : public ConversionPattern {
+  ONNXUniqueOpLowering(TypeConverter &typeConverter, MLIRContext *ctx)
+      : ConversionPattern(
+            typeConverter, mlir::ONNXUniqueOp::getOperationName(), 1, ctx) {}
+
+  ///
+  /// Intermediate data are presented below for better understanding:
+  ///
+  /// there are 4 subtensors sliced along axis 1 of input_x (shape = (2, 4, 2)):
+  /// A: [[1, 1], [1, 1]],
+  ///    [[0, 1], [0, 1]],
+  ///    [[2, 1], [2, 1]],
+  ///    [[0, 1], [0, 1]].
+  ///
+  /// there are 3 unique subtensors:
+  /// [[1, 1], [1, 1]],
+  /// [[0, 1], [0, 1]],
+  /// [[2, 1], [2, 1]].
+  ///
+  /// sorted unique subtensors:
+  /// B: [[0, 1], [0, 1]],
+  ///    [[1, 1], [1, 1]],
+  ///    [[2, 1], [2, 1]].
+  ///
+  /// output_Y is constructed from B:
+  /// [[[0. 1.], [1. 1.], [2. 1.]],
+  /// [[0. 1.], [1. 1.], [2. 1.]]]
+  ///
+  /// output_indices is to map from B to A:
+  /// [1, 0, 2]
+  ///
+  /// output_inverse_indices is to map from A to B:
+  /// [1, 0, 2, 0]
+  ///
+  /// output_counts = [2 1 1]
+  ///
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXTopKOpAdaptor operandAdaptor(operands, op->getAttrDictionary());
+    ONNXUniqueOp uniqueOp = llvm::cast<ONNXUniqueOp>(op);
+    Location loc = op->getLoc();
+    MultiDialectBuilder<KrnlBuilder, IndexExprBuilderForKrnl, MathBuilder,
+        MemRefBuilder>
+        create(rewriter, loc);
+    IndexExprScope scope(create.krnl);
+    ONNXUniqueOpShapeHelper shapeHelper(op, operands, &create.krnlIE);
```

Comment:
Why not call ShapeHelper.computeShape()?

Reply:
Added hapeHelper.computeShape() to calculate the shape here.